File: 2018/9/solution.ts
import * as fs from "fs";
import * as path from "path";

const puzzleInput = fs.readFileSync(path.join(__dirname, "input.txt"), "utf-8");

const [nPlayers, nMarbles] = puzzleInput.match(/\d+/g)!.map(Number);

interface Marble {
  value: number;
  prev?: Marble;
  next?: Marble;
}

class Circle {
  currentMarble: Marble;
  constructor() {
    const marble0: Marble = { value: 0 };
    marble0.next = marble0;
    marble0.prev = marble0;
    this.currentMarble = marble0;
  }

  addMarble(n: number): number {
    if (n % 23 === 0) {
      let curr = this.currentMarble;
      for (let i = 0; i < 7; i++) {
        curr = curr.prev!;
      }
      curr.next!.prev = curr.prev;
      curr.prev!.next = curr.next;
      this.currentMarble = curr.next!;
      return curr.value + n;
    }

    const newMarble: Marble = { value: n };
    const place = this.currentMarble.next!;
    place.next!.prev = newMarble;
    newMarble.next = place.next;
    newMarble.prev = place;
    place.next = newMarble;
    this.currentMarble = newMarble;
    return 0;
  }
}

const circle = new Circle();
const scores = Array(nPlayers).fill(0);
for (let i = 1; i <= nMarbles; i++) {
  scores[(i - 1) % nPlayers] += circle.addMarble(i);
}
console.log(Math.max(...scores));

const circle2 = new Circle();
const scores2 = Array(nPlayers).fill(0);
for (let i = 1; i <= nMarbles * 100; i++) {
  scores2[(i - 1) % nPlayers] += circle2.addMarble(i);
}
console.log(Math.max(...scores2));

File: 2018/9/solution.go
package main

import (
	"container/ring"
	"fmt"
	"os"
	"path/filepath"
	"runtime"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := utils.ExtractInts(string(data))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func newMarble(n int) *ring.Ring {
	marble := ring.New(1)
	marble.Value = n
	return marble
}

func play(nPlayers, lastPoints int) int {
	current := ring.New(1)
	current.Value = 0
	currPlayer := 0
	scores := make(map[int]int, nPlayers)
	for n := range lastPoints {
		if n%23 == 0 {
			scores[currPlayer] += n
			for range 8 {
				current = current.Prev()
			}
			removed := current.Unlink(1)
			scores[currPlayer] += removed.Value.(int)
			current = current.Next()
		} else {
			current = current.Next()
			marble := newMarble(n)
			current.Link(marble)
			current = marble
		}
		currPlayer = (currPlayer + 1) % nPlayers
	}
	res := 0
	for _, score := range scores {
		if score > res {
			res = score
		}
	}
	return res
}

func part1(nums []int) int {
	return play(nums[0], nums[1])
}

func part2(nums []int) int {
	return play(nums[0], nums[1]*100)
}
